src/phoenix/core/traces.py

```diff
@@ -217,20 +240,36 @@ def _consume_spans(self) -> None:
             with self._lock:
                 self._process_span(span)
 
-    def _process_span(self, span: pb.Span) -> None:
+    def _process_span(self, versioned_span: VersionedSpan) -> None:
+        span, version = versioned_span.span, versioned_span.version
         span_id = UUID(bytes=span.context.span_id)
-        existing_span = self._spans.get(span_id)
-        if existing_span and existing_span.end_time:
-            # Reject updates if span has ended.
-            return
+        trace_id = UUID(bytes=span.context.trace_id)
+        if existing_span := self._spans.get(span_id):
+            if version is not None and float(version) < self._max_span_versions[span_id]:
+                # The span could have been previously sequestered for
+                # being orphaned, hence the need for a version check.
+                return
+            existing_trace_id = UUID(bytes=existing_span.context.trace_id)
+            if trace_id != existing_trace_id:
```

Comment:
> Lastly, these backend additions are conceived under very flexible and general terms that are largely independent of llama-index, but these flexibility can be narrowed or reverted entirely depending on what can be changed on the llama-index side.

Not denying that this is conceived under a general term but I think there is a benefit that the client generate the tree - it makes the persistence and collection significantly simpler so we should continue to strive to make that the case. As you pointed out, having IDs change as things get collected has repercussions on the UX as well so I think it's something we should try to avoid.